specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+---
+title: "WAC Specification"
+navTitle: "Spec"
+---
+
+# Specification: WAC
+
+**Status: Prescriptive - Exploratory **
+
+* [Format](#format)
+* [Implementations](#implementations)
+    * [Go](#go)
+    * [Rust/WASM](#rust/wasm)
+
+WAC supports bidirectional transport to and from the [IPLD Data Model].
+In that way it is both a "complete" IPLD Data Model representation and "fitted" to the IPLD Data Model.
+Terminology taken from [Codecs and Completeness].
+
+It takes some inspiration from [Simple DAG], but differs in ways designed to make it reliably round-trip through the IPLD Data Model without loss of fidelity.
+
+The initial motivating use case for WAC is to be a codec that can reliably transport IPLD Data Model data across WebAssembly boundaries while also not being particularly complex to implement.
+
+## Format
+
+### Spec
+
+This format is a series of typing tokens, constant tokens, and inline value data.
+
+Typing tokens are proceeded with the value data for that type.
+
+Every type value can be parsed knowing only the type and without any outside context
+like the container or positional delimiters.
+
+#### Tokens
+
+TODO: Make constants in code match spec
+
+Note: The existing implementations of WAC in https://github.com/aschmahmann/wasm-ipld (Go + Rust/WASM) have different token numbers that we'll need to synchronize before merge.
+
+The tokens for the codec roughly correspond with the [IPLD Data Model Kinds]
+
+They are:
+	Null   = 0
+	True   = 1
+	False  = 2
+	Int    = 3
+	NInt   = 4
+	Float  = 5
+	String = 6
+	Bytes  = 7
+	List   = 8
+	Map    = 9
+	Link   = 10
+
+| Int | Token |
+|---|---|
+| 0 | TYPE_LINK |
+| 1 | TYPE_INTEGER |
+| 2 | TYPE_NEGATIVE_INTEGER |
+| 3 | TYPE_FLOAT |
+| 4 | TYPE_NEGATIVE_FLOAT |
+| 5 | TYPE_STRING |
+| 6 | TYPE_BINARY |
+| 7 | TYPE_MAP |
+| 8 | TYPE_LIST |
+| 9 | VALUE_NULL |
+| 10 | VALUE_TRUE |
+| 11 | VALUE_FALSE |
+
+#### TYPE_LINK
+
+```
+| 0 | CID |
+```
+
+Note: Simple-DAG put a VARINT_LENGTH before the CID indicating how long it was.
+
+CIDs are self-delimiting so this didn't seem necessary.
+
+#### TYPE_INTEGER
+
+```
+| 1 | VARINT |
+```
+
+#### TYPE_SIGNED_INTEGER
+
+```
+| 2 | VARINT |
+```
+
+#### TYPE_FLOAT
+
+```
+| 3 | MATISSA_LENGTH | VARINT
+```
+
+TODO: Floats (and negative floats) need definition here.
+Making implementations actually bidirectional with respect to the IPLD Data Model seems difficult here.
```

Comment:
@rvagg @mikeal @warpfork would like some feedback and thoughts here given that the IPLD Data Model is kind of sketchy on what we mean by floats and what we'd like to be representable.

The text here was grabbed from SimpleDAG, but there wasn't much there to go off of.

Reply:
Yeah, maybe I should write an exploration report about this some time.. we've been iterating toward more accuracy with dag-cbor, although we may have further to go in our specification (our implementations go a bit further than our specifications I think, for reasons).

For this, you could say something like: Double-precision (64-bit) IEEE 754 floats representing finite numbers only. This excludes: half, single or any other length representation, infinites, negative zeros, NaNs, or any other byte representation that does not represent a finite floating point number.

We get pretty good mileage out of being strictly 64-bit IEEE 754 in dag-cbor now, it's been a while since I've seen anything surprising between implementations, although we don't get a ton of float usage by anyone so maybe something will come up.

Reply:
Though, if you're really keen to stick with the mantissa+int thing .. then it might be hard to squish that definition in. If you just stick to 64-bit IEEE 754 then we'd have an easier time with bidirectionality, across languages too since everyone agrees with that binary representation.

I came to the decision that futzing with minimal-length representation thing with floats was a waste of time and too prone to impacting byte format convergence. If you really want floats in your binary data, then 64-bit IEEE 754 gives you what most people are after. If you want anything more then you're going to be doing some weird GPU encoding format optimizing for a strange use-case. And the fact is, people doing content addressing mostly avoid floating point numbers because they suck. (I'd be tempted to even leave it out entirely and let them reconstruct floating point numbers ADL-style).

Reply:
> Though, if you're really keen to stick with the mantissa+int thing

I literally just copied this from https://github.com/mikeal/simple-dag/blob/bd5e533e61c90c392c6b2d05df90c6f1f3843475/README.md?plain=1#L102. No strong opinions here looking for suggestions as to what makes sense.

I'm trying to be close to the "data model" so that the codec is rarely the thing getting in the way of interoperability. However, since the data model is particularly fluffy on float descriptions and comes with a ⚠️ `it is recommended that Float values be avoided when developing systems on IPLD` I'm looking for suggestions.

So looking for advice here. Part of me doesn't like arbitrary size cutoffs (e.g. 64 bits) since I tend to think of the data model as more abstracted than that, but implementations as caring about it. However, I don't think that's something I'm necessarily that attached to if there are better ideas out there.

> (I'd be tempted to even leave it out entirely and let them reconstruct floating point numbers ADL-style).

Something I think would be cool is to basically enable any data that could be represented in an ADL to be representable as a WAC encoded block. Sure, it's unlikely any reasonable implementation would take a 50GB map of maps of maps of bytes and create a single block out of it but being able to do that for 10KB of data sounds useful.

Perhaps it's not reasonable for WAC to reliably round-trip float data through things like ADLs due to the different ways people work with floats.

Reply:
> enable any data that could be represented in an ADL to be representable as a WAC encoded block

I think it'd be nice if this was the case since then an ADL that returns small enough data could be returned as WAC instead of as some other custom format. If that's not doable though then we could probably live with it.

Chatted with @vmx about this a bit as well. I'm currently thinking that even though the IPLD data model doesn't define an explicit range on floats (same as for integers) the added complexity of defining a varint-like spec for floats isn't currently worthwhile given floats aren't highly used in the IPLD ecosystem in general.

As a result, even though we've been able to represent the rest of the data model in WAC (using things like varints), for floats it's reasonable to compromise and use double precision (64-bit) IEEE 754 floats (https://en.wikipedia.org/wiki/Double-precision_floating-point_format).

If it turns out that people need a version of WAC that supports different types of floats then we can always make a WACv2 codec that supports them.

> For this, you could say something like: Double-precision (64-bit) IEEE 754 floats representing finite numbers only. This excludes: half, single or any other length representation, infinites, negative zeros, NaNs, or any other byte representation that does not represent a finite floating point number.

I don't have super strong feelings here it looks like the [WASM component model](https://github.com/WebAssembly/component-model/blob/748a1ddc67c42aff2f9c5884400de61f3b95961b/design/mvp/Explainer.md#type-definitions) folks have settled on a [single canonical NaN](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#loading) value to use but don't say anything about the infinities, negative zero, etc.

Maybe we just say that consumers shouldn't depend on that behavior, but not explicitly cause problems in WAC encoding/decoding since most implementations will probably end up being lazy and use whatever their environment's version of `float64ToBytes/float64FromBytes` anyhow.

Reply:
WDYT about removing the existence of a negative float type further up the document, and having something like:

```suggestion
Floats are represented as big-endian encoded double-precision (64-bit) IEEE 754 floats. Floats MUST represent finite numbers, other values representable as double-precision IEEE 754 floats MAY also be represented. These other values include: infinites, negative zeros, NaNs, or any other byte representation that does not represent a finite floating point number.
```

Note: I went with big-endian here mostly because this seems to be what CBOR uses, if it should be something else lmk 😄

Reply:
Wanted to include a link to the IEEE 754 spec, but everything seems to be behind a paywall. Should I just use https://en.wikipedia.org/wiki/Double-precision_floating-point_format?